esp/services/ws_workunits/ws_workunitsHelpers.cpp

```diff
@@ -1049,25 +1049,81 @@ void WsWuInfo::getInfo(IEspECLWorkunit &info, unsigned long flags)
     getWorkflow(info, flags);
 }
 
+StringBuffer& WsWuInfo::getWorkunitProcessLogPath(const char* process, StringBuffer& path)
+{
+    Owned<IPropertyTreeIterator> procs = cw->getProcesses(process, nullptr);
+    if (!procs->first())
+        return path;
+
+    StringBuffer logSpec;
+    procs->query().getProp("@log", logSpec);
+    if (!logSpec.length())
+        return path;
+
+    splitFilename(logSpec, nullptr, &path, nullptr, nullptr);
+    return path;
+}
+
+unsigned WsWuInfo::getNumberOfSlavesByClusterName()
+{
+    StringAttr clusterName(cw->queryClusterName());
+    if (!clusterName.length()) //Cluster name may not be set yet
+        return 0;
+
+    Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName.str());
+    if (!clusterInfo)
+    {
+        IWARNLOG("Cannot find TargetClusterInfo for workunit %s", cw->queryWuid());
+        return 0;
+    }
+
+    return clusterInfo->getNumberOfSlaveLogs();
+}
+
+unsigned WsWuInfo::getNumberOfSlavesFromZAPFiles()
+{
+    StringBuffer zapWUFolderPath;
+    getWorkunitProcessLogPath("Thor", zapWUFolderPath);
+
+    unsigned numberOfSlaves = 0;
+    Owned<IFile> zapFileFolder = createIFile(zapWUFolderPath);
+    Owned<IDirectoryIterator> thorSlaveFiles = zapFileFolder->directoryFiles("*_thorslave.*");
+    ForEach(*thorSlaveFiles)
+    {
+        const char* pStr = strstr(thorSlaveFiles->query().queryFilename(), "_thorslave.");
+        pStr += 11;
+        const char* slaveNumStrEnd = strchr(pStr, '.');
+        if (!slaveNumStrEnd)
+            continue;
+
+        StringBuffer slaveNumStr;
+        slaveNumStr.append(slaveNumStrEnd - pStr, pStr);
+        unsigned slaveNum = atoi(slaveNumStr);
+        if (slaveNum > numberOfSlaves)
+            numberOfSlaves = slaveNum;
+    }
+    return numberOfSlaves;
+}
+
 unsigned WsWuInfo::getWorkunitThorLogInfo(IArrayOf<IEspECLHelpFile>& helpers, IEspECLWorkunit &info, unsigned long flags, unsigned& helpersCount)
 {
     unsigned countThorLog = 0;
 
     IArrayOf<IConstThorLogInfo> thorLogList;
     if (cw->getWuidVersion() > 0)
     {
-        StringAttr clusterName(cw->queryClusterName());
-        if (!clusterName.length()) //Cluster name may not be set yet
-            return countThorLog;
-
-        Owned<IConstWUClusterInfo> clusterInfo = getTargetClusterInfo(clusterName.str());
-        if (!clusterInfo)
+        unsigned numberOfSlaveLogs = 0;
+        StringAttr thorMasterLogDateSearchString;
+        if (debugImportedStr.length()) //WU imported from a ZAP report
         {
-            IWARNLOG("Cannot find TargetClusterInfo for workunit %s", cw->queryWuid());
-            return countThorLog;
+            numberOfSlaveLogs = getNumberOfSlavesFromZAPFiles();
+            thorMasterLogDateSearchString.set("_thormaster.");
```

Comment:
Isn't there a directory per workunit?
If so, can't pattern be the same, i.e. "thormaster." ?

Reply:
When a ZAP report is created, the '_' is added in the front of the 'thormaster.'

Reply:
ok, you _could_ when you unzip it, normalize it so filenames are same..
I think a directory per workunit with consistently named logs is what we want.
Not sure why ZAP file isn't like that too, any reason it couldn't be?

Reply:
When creating a ZAP report, all of WU files are retrieved/created into one folder with one layer. A WU may have thor logs from >1 thor instances. To identify them, thor instance name is used as a prefix of a thor master log, like: analyt_thor400_90_a_thormaster.2019_02_14

Reply:
right ok, yes that makes sense.
In that case could use a regular pattern in both cases "*thormaster*.log" or similar..
To reduce amount of conditional code?